game/src/sandbox/dashboards/traffic_signals.rs

```diff
@@ -63,41 +59,58 @@ impl TrafficSignalDemand {
             ]))
             .aligned(HorizontalAlignment::Center, VerticalAlignment::Top)
             .build(ctx),
-        })
+        };
+        state.rebuild_world(ctx, app);
+        Box::new(state)
     }
-}
 
-impl State<App> for TrafficSignalDemand {
-    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
-        ctx.canvas_movement();
-        // TODO Use MapspaceTooltips here?
-        if ctx.redo_mouseover() {
-            self.selected = None;
-            app.recalculate_current_selection(ctx);
-            if let Some(ID::Intersection(i)) = app.primary.current_selection.take() {
-                let i = app.primary.map.get_i(i);
-                if i.is_traffic_signal() {
-                    // If we're mousing over something, the cursor is on the map.
-                    let pt = ctx.canvas.get_cursor_in_map_space().unwrap();
-                    for (arrow, count) in self.all_demand[&i.id].make_arrows(i, self.hour) {
-                        if arrow.contains_pt(pt) {
-                            let mut batch = GeomBatch::new();
-                            batch.push(Color::hex("#EE702E"), arrow.clone());
-                            if let Ok(p) = arrow.to_outline(Distance::meters(0.1)) {
-                                batch.push(Color::WHITE, p);
-                            }
-                            let txt = Text::from(format!(
-                                "{} / {}",
-                                prettyprint_usize(count),
-                                self.all_demand[&i.id].count(self.hour).sum()
-                            ));
-                            self.selected = Some((ctx.upload(batch), txt));
-                            break;
-                        }
-                    }
+    fn rebuild_world(&mut self, ctx: &mut EventCtx, app: &App) {
+        let mut world = World::bounded(app.primary.map.get_bounds());
+
+        let mut draw_all = GeomBatch::new();
+        for (i, demand) in &self.all_demand {
+            let cnt_per_movement = demand.count(self.hour);
+            let total_demand = cnt_per_movement.sum();
+
+            let mut outlines = Vec::new();
+            for (movement, cnt) in cnt_per_movement.consume() {
+                let percent = (cnt as f64) / (total_demand as f64);
+                let arrow = app.primary.map.get_i(*i).movements[&movement]
+                    .geom
+                    .make_arrow(percent * Distance::meters(3.0), ArrowCap::Triangle);
+
+                let mut draw_hovered = GeomBatch::new();
+                if let Ok(p) = arrow.to_outline(Distance::meters(0.1)) {
+                    outlines.push(p.clone());
+                    draw_hovered.push(Color::WHITE, p);
                 }
+                draw_all.push(Color::hex("#A3A3A3"), arrow.clone());
+                draw_hovered.push(Color::hex("#EE702E"), arrow.clone());
+
+                world
```

Comment:
Previously, this code was up-front calculating the arrows for every single movement in all intersections. But then when you hover over one intersection, it was lazily calculating those again to render more detailed outlines and tooltips.

Now we just compute everything upfront and keep it in memory.

It could be worth exploring some kind of lazy, hierarchical `World` in the future or if it gets too expensive to do all of this up-front. As in, make a `World` with all of the intersections. Then have some way of lazily populating each movement when you zoom into particular intersections.